library/Booster/Pattern/UnifiedMatcher.hs

```diff
@@ -0,0 +1,831 @@
+                else failWith $ DifferentSorts (Var var) term2
+
+-- unification for lists. Only solves simple cases, returns indeterminate otherwise
+matchLists ::
+    KListDefinition ->
+    [Term] ->
+    Maybe (Term, [Term]) ->
+    [Term] ->
+    Maybe (Term, [Term]) ->
+    StateT MatchState (Except MatchResult) ()
+matchLists
+    def
+    heads1
+    rest1
+    heads2
+    rest2
+        | -- two fully-concrete lists of the same length
+          Nothing <- rest1
+        , Nothing <- rest2 =
+            if length heads1 == length heads2
+                then void $ enqueuePairs heads1 heads2
+                else failWith $ DifferentValues (KList def heads1 rest1) (KList def heads2 rest2)
+        | -- left list has a symbolic part, right one is fully concrete
+          Just (symb1, tails1) <- rest1
+        , Nothing <- rest2 = do
+            let emptyList = KList def [] Nothing
+            remainder <- enqueuePairs heads1 heads2
+            case remainder of
+                Nothing -- equal head length, rest1 must become .List
+                    | null tails1 ->
+                        enqueueRegularProblem symb1 emptyList
+                    | otherwise -> do
+                        -- fully concrete list too short
+                        let surplusLeft = KList def [] rest1
+                        failWith $ DifferentValues surplusLeft emptyList
+                Just (Left leftover1) -> do
+                    -- fully concrete list too short
+                    let surplusLeft = KList def leftover1 rest1
+                    failWith $ DifferentValues surplusLeft emptyList
+                Just (Right leftover2)
+                    | null tails1 -> do
+                        let newRight = KList def leftover2 Nothing
+                        enqueueRegularProblem symb1 newRight
+                    | otherwise -> do
+                        tailRemainder <- -- reversed!
+                            enqueuePairs (reverse tails1) (reverse leftover2)
+                        case tailRemainder of
+                            Nothing ->
+                                -- again symb1 needs to become `.List`
+                                enqueueRegularProblem symb1 emptyList
+                            Just (Left tail1) -> do
+                                -- fully concrete list too short
+                                let surplusLeft = KList def [] $ Just (symb1, reverse tail1)
+                                failWith $ DifferentValues surplusLeft emptyList
+                            Just (Right tail2) -> do
+                                let newRight = KList def (reverse tail2) Nothing
+                                enqueueRegularProblem symb1 newRight
+        | -- mirrored case above: left list fully concrete, right one isn't
+          Nothing <- rest1
+        , Just _ <- rest2 =
+            matchLists def heads2 rest2 heads1 rest1 -- won't loop, will fail later if unification succeeds
```

Comment:
Yeah this was in the original unification code which appeared bidirectional, but we know this would eventually fail. Definitely should refactor, but maybe in a followup?